sql/src/main/java/io/crate/execution/dml/ShardRequestExecutor.java

```diff
@@ -183,7 +191,35 @@ private static ShardId getShardId(ClusterService clusterService,
         ).shardId();
     }
 
-    private static void updateRowCountOrFail(long[] rowCount, ShardResponse response) {
+    private ActionListener<ShardResponse> rowCountListener(RowConsumer consumer, int numberResponse) {
+        return new MultiActionListener<>(
+            numberResponse,
+            () -> new long[]{0},
+            this::countRows,
+            rowCount -> new Row1(rowCount[0]),
+            new OneRowActionListener<>(consumer, Function.identity())
+        );
+    }
+
+    private ActionListener<ShardResponse> collectValuesListener(RowConsumer consumer, int numberResponse) {
+        return new MultiActionListener<>(
+            numberResponse,
+            ArrayList::new,
+            this::collectValues,
+            result -> result.isEmpty() ? Row.EMPTY : new RowN(result.toArray()),
+            new OneRowActionListener<>(consumer, Function.identity())
```

Comment:
Multiple rows may returned here.

Reply:
Yes, good point, this is not correct. This is the wrong type here. My guess is that the result type should be `List<Row>` and not a `Row`, there is also a deduplication step missing I assume. I will come back to this.